File: app/src/main/java/com/abdulrahman/islamicprayertimings/MainActivity.java
package com.abdulrahman.islamicprayertimings;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import javax.net.ssl.HttpsURLConnection;

@SuppressLint("UseSwitchCompatOrMaterialCode")
public class MainActivity extends AppCompatActivity {
    private String country = "Egypt", city = "Cairo";
    private EditText Country, City;
    private Switch sw;
    private boolean S24 = false;
    private Button update;
    private static final String aladhanAPI = "https://api.aladhan.com/v1/timingsByCity?";
    private ExecutorService executorService; // Executor for background tasks
    private Handler mainHandler; // Handler to update the UI thread
    private TextView fajr, shrook, duhr, asr, maghrib, isha;
    private int[] prayerTimingsSeconds;
    private String[] prayerTimings;

    // SharedPreferences name and keys
    private static final String PREFS_NAME = "PrayerAppPrefs";
    private static final String KEY_COUNTRY = "country";
    private static final String KEY_CITY = "city";
    private static final String KEY_FAJR = "fajr";
    private static final String KEY_SHROOK = "shrook";
    private static final String KEY_DUHR = "duhr";
    private static final String KEY_ASR = "asr";
    private static final String KEY_MAGHRIB = "maghrib";
    private static final String KEY_ISHA = "isha";

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main_activity);
        initUI();
        // Initialize ExecutorService and Handler
        executorService = Executors.newSingleThreadExecutor(); // For single background task
        mainHandler = new Handler(Looper.getMainLooper()); // For updating UI from the background thread

        // Get permission for notification
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            // Check if the app has the POST_NOTIFICATIONS permission
            if (ContextCompat.checkSelfPermission(this, android.Manifest.permission.POST_NOTIFICATIONS)
                    != PackageManager.PERMISSION_GRANTED) {
                // Request the permission if not granted
                ActivityCompat.requestPermissions(this,
                        new String[]{android.Manifest.permission.POST_NOTIFICATIONS},
                        1);
            }
        }

        // Load saved country and city from SharedPreferences
        loadPreferences();
        setupListeners();
        getTimings();
    }

    private void initUI() {
        fajr = findViewById(R.id.fajr);
        shrook = findViewById(R.id.shrook);
        duhr = findViewById(R.id.duhr);
        asr = findViewById(R.id.asr);
        maghrib = findViewById(R.id.maghrib);
        isha = findViewById(R.id.isha);
        Country = findViewById(R.id.Country);
        City = findViewById(R.id.City);
        update = findViewById(R.id.button);
        sw = findViewById(R.id.switch1);
    }

    private void setupListeners() {
        update.setOnClickListener(view -> {
            // Save the entered values when the update button is clicked
            savePreferences();
            getTimings();
        });
        // Set an OnCheckedChangeListener to handle the switch state change
        sw.setOnCheckedChangeListener((buttonView, isChecked) -> {
            S24 = isChecked;
            SharedPreferences sharedPreferences = getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
            SharedPreferences.Editor editor = sharedPreferences.edit();
            editor.putBoolean("time_format", S24);
            editor.apply();
        });
    }

    private void savePreferences() {
        // Get the values entered by the user
        country = Country.getText().toString().trim();
        city = City.getText().toString().trim();

        // Store the values in SharedPreferences
        SharedPreferences sharedPreferences = getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString(KEY_COUNTRY, country);
        editor.putString(KEY_CITY, city);
        editor.putBoolean("time_format", S24);
        editor.apply();
    }

    private void loadPreferences() {
        // Retrieve the stored values from SharedPreferences
        SharedPreferences sharedPreferences = getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
        country = sharedPreferences.getString(KEY_COUNTRY, "Egypt");
        city = sharedPreferences.getString(KEY_CITY, "Cairo");
        Country.setText(country);
        City.setText(city);
        S24 = sharedPreferences.getBoolean("time_format", false);
        sw.setChecked(S24);

        // Load prayer timings
        fajr.setText(sharedPreferences.getString(KEY_FAJR, ""));
        shrook.setText(sharedPreferences.getString(KEY_SHROOK, ""));
        duhr.setText(sharedPreferences.getString(KEY_DUHR, ""));
        asr.setText(sharedPreferences.getString(KEY_ASR, ""));
        maghrib.setText(sharedPreferences.getString(KEY_MAGHRIB, ""));
        isha.setText(sharedPreferences.getString(KEY_ISHA, ""));
    }

    private void savePrayerTimings(String fajrTime, String shrookTime, String duhrTime,
                                   String asrTime, String maghribTime, String ishaTime) {
        SharedPreferences sharedPreferences = getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString(KEY_FAJR, fajrTime);
        editor.putString(KEY_SHROOK, shrookTime);
        editor.putString(KEY_DUHR, duhrTime);
        editor.putString(KEY_ASR, asrTime);
        editor.putString(KEY_MAGHRIB, maghribTime);
        editor.putString(KEY_ISHA, ishaTime);
        editor.apply();
    }

    private void showToast(String message) {
        runOnUiThread(() -> Toast.makeText(this, message, Toast.LENGTH_SHORT).show());
    }
    private String convert24HourTo12Hour(String time24) {
        // Check if the input time is valid
        if (time24 == null || !time24.matches("([01]?[0-9]|2[0-3]):[0-5][0-9]")) {
            throw new IllegalArgumentException("Invalid 24-hour time format");
        }
        // Split the input string into hour and minute
        String[] timeParts = time24.split(":");
        int hour = Integer.parseInt(timeParts[0]);
        String minute = timeParts[1];
        // Determine AM or PM
        String period = hour < 12 ? "AM" : "PM";
        // Convert the hour to 12-hour format
        if (hour == 0) {
            hour = 12; // Midnight case
        } else if (hour > 12) {
            hour -= 12; // Convert hours greater than 12 to 12-hour format
        }
        // Format the time in 12-hour format with AM/PM
        return hour + ":" + minute + " " + period;
    }
    private void getTimings() {
        executorService.execute(() -> {
            try {
                // Use the loaded country and city
                URL url = new URL(aladhanAPI + "city=" + city + "&country=" + country);
                HttpsURLConnection conn = (HttpsURLConnection) url.openConnection();
                conn.setRequestMethod("GET");

                if (conn.getResponseCode() != HttpURLConnection.HTTP_OK) {
                    showToast(conn.getResponseMessage());
                    return;
                }
                // Read the response
                BufferedReader in = new BufferedReader(new InputStreamReader(conn.getInputStream()));
                StringBuilder response = new StringBuilder();
                String line;
                while ((line = in.readLine()) != null) {
                    response.append(line);
                }
                in.close();
                // Parse the response JSON
                JSONObject jsonResponse = new JSONObject(response.toString());
                JSONObject data = jsonResponse.getJSONObject("data");
                JSONObject timings = data.getJSONObject("timings");

                String Fajr = timings.getString("Fajr");
                String Shrook = timings.getString("Sunrise");
                String Duhr = timings.getString("Dhuhr");
                String Asr = timings.getString("Asr");
                String Maghrib = timings.getString("Maghrib");
                String Isha = timings.getString("Isha");

                prayerTimings = new String[6];
                if(!S24){
                    prayerTimings[0] = convert24HourTo12Hour(Fajr);
                    prayerTimings[1] = convert24HourTo12Hour(Shrook);
                    prayerTimings[2] = convert24HourTo12Hour(Duhr);
                    prayerTimings[3] = convert24HourTo12Hour(Asr);
                    prayerTimings[4] = convert24HourTo12Hour(Maghrib);
                    prayerTimings[5] = convert24HourTo12Hour(Isha);
                }
                else{
                    prayerTimings[0] = Fajr;
                    prayerTimings[1] = Shrook;
                    prayerTimings[2] = Duhr;
                    prayerTimings[3] = Asr;
                    prayerTimings[4] = Maghrib;
                    prayerTimings[5] = Isha;
                }

                // Save prayer timings to SharedPreferences
                savePrayerTimings(prayerTimings[0], prayerTimings[1], prayerTimings[2], prayerTimings[3], prayerTimings[4], prayerTimings[5]);

                mainHandler.post(() -> {
                    String[] temp = {Fajr, Shrook, Duhr, Asr, Maghrib, Isha};
                    prayerTimingsSeconds = new int[6];
                    for (int i = 0; i < 6; i++) {
                        String[] timeParts = temp[i].split(":");
                        int hours = Integer.parseInt(timeParts[0]);
                        int minutes = Integer.parseInt(timeParts[1]);
                        int totalSeconds = (hours * 3600) + (minutes * 60);
                        prayerTimingsSeconds[i] = totalSeconds;
                    }

                    fajr.setText(prayerTimings[0]);
                    shrook.setText(prayerTimings[1]);
                    duhr.setText(prayerTimings[2]);
                    asr.setText(prayerTimings[3]);
                    maghrib.setText(prayerTimings[4]);
                    isha.setText(prayerTimings[5]);

                    Intent serviceIntent = new Intent(this, NextService.class);
                    serviceIntent.putExtra("prayer_timings_seconds", prayerTimingsSeconds);
                    serviceIntent.putExtra("prayer_timings", prayerTimings);
                    startService(serviceIntent);
                });
            } catch (IOException e) {
                showToast("Error getting timings from API");
            } catch (JSONException e) {
                showToast("Error parsing JSON response");
            }
        });
    }
}

File: app/src/main/java/com/example/islamicprayertimings/MainActivity.kt
package com.example.islamicprayertimings

import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SwitchCompat
import androidx.core.content.edit
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import okhttp3.Call
import okhttp3.Response
import org.json.JSONObject
import java.io.IOException

class MainActivity : AppCompatActivity(){
  private val aladhanAPI = "https://api.aladhan.com/v1/timingsByCity?"
  private lateinit var swipeRefreshLayout: SwipeRefreshLayout

  private lateinit var country : TextView
  private lateinit var city : TextView

  private lateinit var fajr : TextView
  private lateinit var sunrise : TextView
  private lateinit var dhuhr : TextView
  private lateinit var asr : TextView
  private lateinit var maghrib : TextView
  private lateinit var isha : TextView

  private lateinit var updateBtn : Button
  private lateinit var switch24h: SwitchCompat
  private lateinit var switchSound: SwitchCompat


  private var prayerTimings24 = Array<String>(6) {""}
  private var prayerTimings12 = Array<String>(6) {""}
  private var prayerTimingsInSeconds = IntArray(6)

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.main_activity)

    refUI()
    setupListeners()
    getNotificationsPerm()
    loadSharedPreferences()
    getTimings()
  }

  override fun onDestroy() {
    super.onDestroy()
    saveSharedPreferences()
  }

  private fun refUI(){
    swipeRefreshLayout = findViewById(R.id.swipeRefreshLayout)

    country = findViewById<TextView>(R.id.Country)
    city = findViewById<TextView>(R.id.City)

    fajr = findViewById<TextView>(R.id.fajr)
    sunrise = findViewById<TextView>(R.id.sunrise)
    dhuhr = findViewById<TextView>(R.id.dhuhr)
    asr = findViewById<TextView>(R.id.asr)
    maghrib = findViewById<TextView>(R.id.maghrib)
    isha = findViewById<TextView>(R.id.isha)

    updateBtn = findViewById<Button>(R.id.button)
    switch24h = findViewById<SwitchCompat>(R.id.switch_24hour)
    switchSound = findViewById<SwitchCompat>(R.id.switch_sound)
  }

  private fun getNotificationsPerm(){
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
      val permission = android.Manifest.permission.POST_NOTIFICATIONS
      if (checkSelfPermission(permission) != PackageManager.PERMISSION_GRANTED) {
        requestPermissions(arrayOf(permission), 100)
      }
    }
  }

  private fun setupListeners() {
    swipeRefreshLayout.setOnRefreshListener {
      getTimings()
    }
    updateBtn.setOnClickListener {
      swipeRefreshLayout.isRefreshing = true
      getTimings()
    }
    switch24h.setOnCheckedChangeListener {_, isChecked ->
      if (isChecked) {
        fajr.text = prayerTimings24[0]
        sunrise.text = prayerTimings24[1]
        dhuhr.text = prayerTimings24[2]
        asr.text = prayerTimings24[3]
        maghrib.text = prayerTimings24[4]
        isha.text = prayerTimings24[5]
      } else {
        fajr.text = prayerTimings12[0]
        sunrise.text = prayerTimings12[1]
        dhuhr.text = prayerTimings12[2]
        asr.text = prayerTimings12[3]
        maghrib.text = prayerTimings12[4]
        isha.text = prayerTimings12[5]
      }
      startNextPrayService()
    }
    switchSound.setOnCheckedChangeListener { _,isChecked ->
      startNextPrayService()
    }
  }

  private fun saveSharedPreferences() {
    val sharedPreferences = getSharedPreferences("prefs", MODE_PRIVATE)
    sharedPreferences.edit {
      putString("country", country.text.toString().trim())
      putString("city", city.text.toString().trim())

      putString("fajr", prayerTimings24[0])
      putString("sunrise", prayerTimings24[1])
      putString("dhuhr", prayerTimings24[2])
      putString("asr", prayerTimings24[3])
      putString("maghrib", prayerTimings24[4])
      putString("isha", prayerTimings24[5])

      putBoolean("s24", switch24h.isChecked)
      putBoolean("s_sound", switchSound.isChecked)
    }
  }

  private fun loadSharedPreferences(){
    val sharedPreferences = getSharedPreferences("prefs", MODE_PRIVATE)
    country.text = sharedPreferences.getString("country", "Egypt")
    city.text = sharedPreferences.getString("city", "Cairo")

    prayerTimings24[0] = sharedPreferences.getString("fajr", "").toString()
    prayerTimings24[1] = sharedPreferences.getString("sunrise", "").toString()
    prayerTimings24[2] = sharedPreferences.getString("dhuhr", "").toString()
    prayerTimings24[3] = sharedPreferences.getString("asr", "").toString()
    prayerTimings24[4] = sharedPreferences.getString("maghrib", "").toString()
    prayerTimings24[5] = sharedPreferences.getString("isha", "").toString()

    for (i in 0 until  6) if (prayerTimings24[0].isNotEmpty()) {
      prayerTimings12[i] = convertTo12HourFormat(prayerTimings24[i])
      val parts = prayerTimings24[i].split(":")
      prayerTimingsInSeconds[i] = parts[0].toInt() * 3600 + parts[1].toInt() * 60
    }

    fajr.text = prayerTimings24[0]
    sunrise.text = prayerTimings24[1]
    dhuhr.text = prayerTimings24[2]
    asr.text = prayerTimings24[3]
    maghrib.text = prayerTimings24[4]
    isha.text = prayerTimings24[5]

    switch24h.isChecked = sharedPreferences.getBoolean("s24", true)
    switchSound.isChecked = sharedPreferences.getBoolean("s_sound", false)
  }

  private fun getTimings(){
    val client = okhttp3.OkHttpClient()
    val request = okhttp3.Request.Builder().url(aladhanAPI + "city=" + city.text.trim() +
            "&country=" + country.text.trim()).build()
    client.newCall(request).enqueue(object : okhttp3.Callback{
      override fun onFailure(call: Call, e: IOException) {
        runOnUiThread {
          Toast.makeText(this@MainActivity, "No internet connection", Toast.LENGTH_SHORT).show()
          swipeRefreshLayout.isRefreshing = false
        }
      }

      override fun onResponse(call: Call, response: Response) {
        response.use {
          if (!response.isSuccessful) {
            runOnUiThread {
              Toast.makeText(this@MainActivity, "Failed to retrieve prayer timings", Toast.LENGTH_LONG).show()
              swipeRefreshLayout.isRefreshing = false
            }
            return
          }
          val body = response.body?.string()
          if (body.isNullOrEmpty()) return

          try{
            val jsonResponse = JSONObject(body)
            val data = jsonResponse.getJSONObject("data")
            val timings = data.getJSONObject("timings")

            prayerTimings24[0] = toDefaultLocale(timings.getString("Fajr"))
            prayerTimings24[1] = toDefaultLocale(timings.getString("Sunrise"))
            prayerTimings24[2] = toDefaultLocale(timings.getString("Dhuhr"))
            prayerTimings24[3] = toDefaultLocale(timings.getString("Asr"))
            prayerTimings24[4] = toDefaultLocale(timings.getString("Maghrib"))
            prayerTimings24[5] = toDefaultLocale(timings.getString("Isha"))

            for (i in 0 until  6){
              prayerTimings12[i] = convertTo12HourFormat(prayerTimings24[i])

              val parts = prayerTimings24[i].split(":")
              prayerTimingsInSeconds[i] = parts[0].toInt() * 3600 + parts[1].toInt() * 60
            }

            runOnUiThread {
              if(switch24h.isChecked){
                fajr.text = prayerTimings24[0]
                sunrise.text = prayerTimings24[1]
                dhuhr.text = prayerTimings24[2]
                asr.text = prayerTimings24[3]
                maghrib.text = prayerTimings24[4]
                isha.text = prayerTimings24[5]
              }
              else{
                fajr.text = prayerTimings12[0]
                sunrise.text = prayerTimings12[1]
                dhuhr.text = prayerTimings12[2]
                asr.text = prayerTimings12[3]
                maghrib.text = prayerTimings12[4]
                isha.text = prayerTimings12[5]
              }
              swipeRefreshLayout.isRefreshing = false
              saveSharedPreferences()
            }
          } catch (_: Exception) {
            runOnUiThread {
              Toast.makeText(this@MainActivity, "Error parsing response", Toast.LENGTH_LONG).show()
            }
          }
        }
      }
    })
    startNextPrayService()
  }

  fun convertTo12HourFormat(time24: String): String {
    if (time24.isBlank()) return ""
    return try {
      val inputFormat = java.text.SimpleDateFormat("HH:mm", java.util.Locale.getDefault())
      val outputFormat = java.text.SimpleDateFormat("h:mm a", java.util.Locale.getDefault())
      val date = inputFormat.parse(time24)
      outputFormat.format(date!!)
    } catch (_: Exception) {
      ""
    }
  }

  fun toDefaultLocale(time24: String): String {
    if (time24.isBlank()) return ""
    return try{
      val inputFormat = java.text.SimpleDateFormat("HH:mm", java.util.Locale.getDefault())
      val outputFormat = java.text.SimpleDateFormat("HH:mm", java.util.Locale.getDefault())
      val date = inputFormat.parse(time24)
      outputFormat.format(date!!)
    } catch (_: Exception) {
      ""
    }
  }
  fun startNextPrayService(){
    if(prayerTimings24[0] != ""){
      val serviceIntent = Intent(this@MainActivity, NextPrayService::class.java).apply {
        if(switch24h.isChecked){
          putExtra("PrayerTimings", prayerTimings24)
        }
        else{
          putExtra("PrayerTimings", prayerTimings12)
        }
        putExtra("prayerTimingsInSeconds", prayerTimingsInSeconds)
        putExtra("adhanSound", switchSound.isChecked)
      }
      startService(serviceIntent)
    }
  }
}